File: Hard/668. Kth Smallest Number in Multiplication Table/668. Kth Smallest Number in Multiplication Table.java
class Solution {
    public int findKthNumber(int m, int n, int k) {
        int left = 1;
        int right = m * n;

        while (left < right) {
            int mid = left + (right - left) / 2;
            
            // Count the number of elements less than or equal to mid in the multiplication table
            int count = countElements(m, n, mid);
            
            if (count < k) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        
        return left;
    }
    
    private int countElements(int m, int n, int target) {
        int count = 0;
        for (int i = 1; i <= m; i++) {
            count += Math.min(target / i, n);
        }
        return count;
    }
}

File: Hard/668. Kth Smallest Number in Multiplication Table/668. Kth Smallest Number in Multiplication Table.py
class Solution:
    def findKthNumber(self, m: int, n: int, k: int) -> int:
        left, right = 1, m * n

        while left < right:
            mid = left + (right - left) // 2

            # Count the number of elements less than or equal to mid in the multiplication table
            count = self.countElements(m, n, mid)

            if count < k:
                left = mid + 1
            else:
                right = mid

        return left

    def countElements(self, m: int, n: int, target: int) -> int:
        count = 0
        for i in range(1, m + 1):
            count += min(target // i, n)
        return count
